### ISSUE TITLE
[QUERY] Java Azure Async Client for ESB: Parallel vs Sequential read

### ISSUE THREAD

**Issue by wernerdaehn** [Azure/azure-sdk-for-java#18730]
**Query/Question**
I would like to get the messages from a single queue **sequentially** using the async client.
During debugging I can see that there are 8 threads listening one the queue. Also when I set a breakpoint on the message handler other messages are read meanwhile.
But I need a true sequential processing: Wait for a message, process it, release it from the queue, pickup the next message.

Can anybody spot my error in below code?

***Why is this not a Bug or a feature Request?***
I have read the documentation extensively, the docs, the Java samples and also Reactive programming but have not found any solution. Hence I assume the problem is on my side, you might actually consider adding documentation in that regards. Should be a typical question.

**Setup (please complete the following information if applicable):**
 - OS: Linux
 - IDE : Eclipse
 - azure-storage-file-datalake: v12.2.0
 - azure-messaging-servicebus: v7.0.0

			receiver = new ServiceBusClientBuilder()
				    .connectionString(servicebusconnectionstring)
				    .receiver()
				    .queueName(queuename)
				    .buildAsyncClient();
			subscription = receiver
					.receiveMessages()
					.map(message -> {
					    byte[] payload = message.getBody().toBytes();
					    return handleMessage(payload, message);
					})
					.subscribe(aVoid -> {
						}, error -> {
							sendError(error);
						}, () -> {
							logger.info("Azure queue reader is shut down");
						});

**Comment by hemanttanwar** [comment]
There are few things to consider while designing your application. 

> If you are needing absolute guarantee in maintaining the order, consider session queue. Here is more detail about it https://docs.microsoft.com/en-us/azure/service-bus-messaging/message-sessions  

Sample : https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/servicebus/azure-messaging-servicebus/src/samples/java/com/azure/messaging/servicebus/ServiceBusSessionProcessorSample.java
Our processor are using sync receiver in background.

When you have async receiver client, like in your code above, keep in mind that. Service does not stop from creating more clients from other applications or other part of your code. So multiple clients could be reading from same queue. Session queue is solution where you can ensure that only once client can read at one point in time.

Using Project Reactor  API and and async receiver client, you may consider using `concatMap` operator, which  sequentially and preserving order https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#concatMap-java.util.function.Function-   but this can not stop other clients from reading messages from the queue.

I do not see any issue in your code above, you are using `map()`. I did try with 100 messages in a queue with one async client client and I am receiving in order. Where do you see 8 thread listening ? Please provide more details.

**Comment by wernerdaehn** [comment]
Okay. Your post caused me trying to create the most simple test and in essence you are right, there is no parallel processing. I omitted the essential part: The handlemessage() method is using an async file reader to read the file.
So what happens is that a message is received, the handlemessage starts the file read and thus is finished right away, the next message is read. So when I set a breakpoint inside the async file reader, then multiple messages are processed.

I'll close this request and create a new one.

Thanks for your help, sorry for my mistake!
